### PR TITLE
VR- teleop support 

### PR DESCRIPTION
**Pull request by liuyulinn**
1. add save or not option in recording
2. bugfix replay 
3. add two dexterity envorinment [need to change assets path]

### CODE REVIEW

**Comment by StoneT2000** [comment]
Can you merge in the main branch? there are some conflicts. Just override with the code from main, (e.g. so100 which you didn't modify at all).

File: mani_skill/envs/tasks/dexterity/open_notebook.py [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+        robot_init_qpos_noise=0.02,
+        # obj_init_pos_noise=0.02,
+        # difficulty_level: int = -1,
+        num_envs=1,
+        reconfiguration_freq=None,
+        **kwargs,
+    ):
+        self.robot_init_qpos_noise = robot_init_qpos_noise
+        # self.obj_init_pos_noise = obj_init_pos_noise
+        # self.obj_heights: torch.Tensor = torch.Tensor()
```
Comment (StoneT2000) [inline mani_skill/envs/tasks/dexterity/open_notebook.py]: are these commented lines needed? Lets remove them and clean these files up a bit. We would like to have nice well written dexterous hand tasks.

File: mani_skill/agents/robots/ur_e/ur_10e.py [hunk 2]
```diff
@@ -43,3 +45,31 @@ def _controller_configs(
                 use_delta=True,
             ),
         )
+
```
Comment (StoneT2000) [inline mani_skill/agents/robots/ur_e/ur_10e.py]: do we need these commented lines?

File: mani_skill/agents/robots/ur_e/ur_10e.py [hunk 3]
```diff
@@ -1,11 +1,13 @@
 import numpy as np
 import sapien
 
+import torch 
 from mani_skill import ASSET_DIR, PACKAGE_ASSET_DIR
 from mani_skill.agents.base_agent import BaseAgent, Keyframe
 from mani_skill.agents.controllers import *
 from mani_skill.agents.registration import register_agent
-
+from mani_skill.utils.structs.actor import Actor
```
Comment (StoneT2000) [inline mani_skill/agents/robots/ur_e/ur_10e.py]: are these imports needed?

File: mani_skill/envs/sapien_env.py [hunk 4]
```diff
@@ -1069,10 +1072,16 @@ def _step_action(
                 action = common.batch(action)
             self.agent.set_action(action)
             if self._sim_device.is_cuda():
-                if self.agent.controller.sets_target_qpos:
+                if isinstance(self.agent.controller, dict):
+                    # TODO: a small optimization is to cache whether the dict of controllers has any that set qpos/qvel values
```
Comment (StoneT2000) [inline mani_skill/envs/sapien_env.py]: is this my comment? How did that end up here

File: mani_skill/envs/tasks/dexterity/insert_flower.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+from mani_skill.utils.structs.actor import Actor
+from mani_skill.utils.structs.pose import Pose, vectorize_pose
+from mani_skill.utils.structs.types import Array, GPUMemoryConfig, SimConfig
+
+
+@register_env(
+    "InsertFlower-v1",
+    max_episode_steps=300,
+)
+class InsertFlowerEnv(BaseEnv):
```
Comment (StoneT2000) [inline mani_skill/envs/tasks/dexterity/insert_flower.py:32]: nice to have more good dextrous hand tasks!

File: mani_skill/envs/tasks/dexterity/insert_flower.py [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+        num_envs=1,
+        reconfiguration_freq=None,
+        **kwargs,
+    ):
+        self.robot_init_qpos_noise = robot_init_qpos_noise
+        self.target_area = {"min": [-0.3, -0.25, 0.25], "max": [-0.2, -0.15, 0.35]}
+
+        super().__init__(
+            *args,
+            # robot_uids="allegro_hand_right_floating",
```
Comment (StoneT2000) [inline mani_skill/envs/tasks/dexterity/insert_flower.py]: how come robot uid is not set?

File: mani_skill/envs/tasks/dexterity/insert_flower.py [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+
+        # flower_builder.set_material(flower_material)
+        # Pose([-0.269485, 0.0198321, 0.016], )
+        self.init_flower_pose = Pose.create_from_pq(
+            [-0.242, 0.0, 0.015 + 0.001], [-0.352413, -0.258145, -0.635074, 0.637062]
+        )
+        flower_builder.initial_pose = self.init_flower_pose
+        self.flower = flower_builder.build(name="flower")
+
+        # for i, obj in enumerate(self.flower._objs):
```
Comment (StoneT2000) [inline mani_skill/envs/tasks/dexterity/insert_flower.py]: are these commented lines needed? Lets remove them and clean these files up a bit. We would like to have nice well written dexterous hand tasks.

File: mani_skill/envs/tasks/dexterity/insert_flower.py [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+        # Check if the object is within the specified bounds
+        is_within = torch.logical_and(
+            torch.all(object_pos >= self.target_area_box[0], dim=-1),  # min bounds
+            torch.all(object_pos <= self.target_area_box[1], dim=-1),  # max bounds
+        )
+
+        return {"success": is_within}
+
+    def compute_dense_reward(self, obs: Any, action: Array, info: Dict) -> float:
+        # return 0
```
Comment (StoneT2000) [inline mani_skill/envs/tasks/dexterity/insert_flower.py]: remove this comment

File: mani_skill/envs/tasks/dexterity/insert_flower.py [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+            torch.max(object_pos - self.target_area_box[1], torch.zeros_like(object_pos)),  # upper bound
+        )
+        reward = torch.exp(-5 * torch.norm(dist_outside)).reshape(-1)
+
+        return reward
+
+    def compute_normalized_dense_reward(self, obs: Any, action: Array, info: Dict):
+        # pass
+        # this should be equal to compute_dense_reward / max possible reward
+        return self.compute_dense_reward(obs=obs, action=action, info=info) / 4.0
```
Comment (StoneT2000) [inline mani_skill/envs/tasks/dexterity/insert_flower.py:185]: have we verified these reward functions?

e.g. have we tried running PPO/SAC?

File: mani_skill/trajectory/replay_trajectory.py [hunk 10]
```diff
@@ -129,9 +126,9 @@ def replay_parallelized_sim(
         trajectory_ids = [episode["episode_id"] for episode in episode_batch]
         episode_lens = np.array([episode["elapsed_steps"] for episode in episode_batch])
         ori_control_mode = episode_batch[0]["control_mode"]
-        assert all(
-            [episode["control_mode"] == ori_control_mode for episode in episode_batch]
-        ), "Replay trajectory with parallelized environments is only supported for trajectories with the same control mode"
+        assert all([episode["control_mode"] == ori_control_mode for episode in episode_batch]), (
```
Comment (StoneT2000) [inline mani_skill/trajectory/replay_trajectory.py:130]: why are these changed? formatter?

File: mani_skill/utils/scene_builder/table/scene_builder.py [hunk 11]
```diff
@@ -282,3 +282,16 @@ def initialize(self, env_idx: torch.Tensor):
             self.env.agent.robot.set_pose(
                 sapien.Pose([-0.725, 0, 0], q=euler2quat(0, 0, np.pi / 2))
             )
+        
+        elif self.env.robot_uids == "ur_10e":
+            # qpos = np.array([0, np.pi / 2, np.pi / 2, np.pi / 2, -np.pi / 2, 1.0])
```
Comment (StoneT2000) [inline mani_skill/utils/scene_builder/table/scene_builder.py]: if not needed lets remove

File: mani_skill/utils/wrappers/record.py [hunk 12]
```diff
@@ -358,20 +357,18 @@ def reset(
...
         seed: Optional[Union[int, List[int]]] = None,
         options: Optional[dict] = dict(),
+        save=True,
         **kwargs,
     ):
-
         if self.save_on_reset:
             if self.save_video and self.num_envs == 1:
-                self.flush_video()
+                self.flush_video(save=save)
```
Comment (StoneT2000) [inline mani_skill/utils/wrappers/record.py:365]: what's the reasoning for having this save option in the reset function?

Couldn't we just at wrapper creation pass save_on_reset=False?

**Review by StoneT2000 — CHANGES_REQUESTED** [review]

File: mani_skill/envs/tasks/dexterity/insert_flower.py [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+from mani_skill.utils.registration import register_env
+from mani_skill.utils.scene_builder.table import TableSceneBuilder
+from mani_skill.utils.structs.actor import Actor
+from mani_skill.utils.structs.pose import Pose, vectorize_pose
+from mani_skill.utils.structs.types import Array, GPUMemoryConfig, SimConfig
+
+
+@register_env(
+    "InsertFlower-v1",
+    max_episode_steps=300,
```
Comment (StoneT2000) [inline mani_skill/envs/tasks/dexterity/insert_flower.py:29]: I think it would be good to include the oakink-v2 assets. We usually define new assets here: https://github.com/haosulab/ManiSkill/blob/main/mani_skill/utils/assets/data.py

Can you add a new data group for that? You can follow how e.g. the scene datasets are defined which are hosted on hugging face. We expect a zip file with all the data we need to extract into a folder.

Finally after doing so to ensure users don't run this env without downloading assets first, can you add a `asset_download_ids=[asset_id])` to the register_env function? 

https://github.com/haosulab/ManiSkill/blob/main/mani_skill/envs/tasks/tabletop/pick_single_ycb.py#L26C57-L26C84 is a example of how it is done.

**Review by StoneT2000 — CHANGES_REQUESTED** [review]

File: mani_skill/utils/wrappers/record.py (hunk 12 above)
Comment (liuyulinn) [reply mani_skill/utils/wrappers/record.py:365]: The current implementation is: start to record when stepping actions, user can decide whether to keep or throw it at the end. And also we record multiple trajs with different keep & throw options. If set save_on_reset = False, then it need to manually set env.save_on_reset = False & env.save_on_reset = True everytime and I don't think it's a good way to set attribution instead of parsing args.

File: mani_skill/utils/scene_builder/table/scene_builder.py (hunk 11 above)
Comment (liuyulinn) [reply mani_skill/utils/scene_builder/table/scene_builder.py]: OK. I will make another merge request.

File: mani_skill/trajectory/replay_trajectory.py (hunk 10 above)
Comment (liuyulinn) [reply mani_skill/trajectory/replay_trajectory.py:130]: Yes just formatter. By ruff.

File: mani_skill/envs/tasks/dexterity/insert_flower.py (hunk 9 above)
Comment (liuyulinn) [reply mani_skill/envs/tasks/dexterity/insert_flower.py:185]: reward function not yet verfied.

File: mani_skill/envs/tasks/dexterity/insert_flower.py (hunk 8 above)
Comment (liuyulinn) [reply mani_skill/envs/tasks/dexterity/insert_flower.py]: will do that.

File: mani_skill/envs/tasks/dexterity/insert_flower.py (hunk 6 above)
Comment (liuyulinn) [reply mani_skill/envs/tasks/dexterity/insert_flower.py]: Will set a default one.

File: mani_skill/envs/tasks/dexterity/insert_flower.py (hunk 7 above)
Comment (liuyulinn) [reply mani_skill/envs/tasks/dexterity/insert_flower.py]: remove all.

File: mani_skill/agents/robots/ur_e/ur_10e.py (hunk 2 above)
Comment (liuyulinn) [reply mani_skill/agents/robots/ur_e/ur_10e.py]: I will just overload the ur_10e.py file since it's not actually modified.

File: mani_skill/envs/tasks/dexterity/open_notebook.py (hunk 1 above)
Comment (liuyulinn) [reply mani_skill/envs/tasks/dexterity/open_notebook.py]: I will update the open_notebook env later after finishing the success condition & reward.
